Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm

```diff
@@ -662,18 +662,47 @@ static void registerLogHook()
     os_log_set_hook(OS_LOG_TYPE_DEFAULT, ^(os_log_type_t type, os_log_message_t msg) {
         if (msg->buffer_sz > 1024)
             return;
-        char* messageString = os_log_copy_message_string(msg);
-        String logString = String::fromUTF8(messageString);
-        free(messageString);
-
-        String logChannel = String::fromUTF8(msg->subsystem);
-        String logCategory = String::fromUTF8(msg->category);
-
-        callOnMainRunLoop([logChannel = logChannel.isolatedCopy(), logCategory = logCategory.isolatedCopy(), logString = logString.isolatedCopy(), type] {
-            auto* connection = WebProcess::singleton().existingNetworkProcessConnection();
-            if (!connection)
-                return;
-            connection->connection().send(Messages::NetworkConnectionToWebProcess::LogOnBehalfOfWebContent(logChannel, logCategory, logString, type, getpid()), 0);
+
+        Vector<char> logFormat(msg->format, msg->format ? strlen(msg->format) + 1 : 0);
+        Vector<char> logChannel(msg->subsystem, msg->subsystem ? strlen(msg->subsystem) + 1 : 0);
+        Vector<char> logCategory(msg->category, msg->category ? strlen(msg->category) + 1 : 0);
+
+        Vector<uint8_t> buffer(msg->buffer, msg->buffer_sz);
+        Vector<uint8_t> privdata(msg->privdata, msg->privdata_sz);
+
+        static NeverDestroyed<Ref<WorkQueue>> queue(WorkQueue::create("Log Queue", WorkQueue::QOS::Background));
+
+        NetworkProcessConnection* connection = nullptr;
+        RefPtr<IPC::Connection> ipcConnection;
+        if (isMainThread())
```

Comment:
Is this ever called on the main thread?

Reply:
> Is this ever called on the main thread?

Yes, that is the normal case. The hook will be called on the same thread as the logging took place, and most logging is being submitted from the main thread.

Reply:
> Here are my thoughts:
> 
> 1. If this never gets called on the main thread, then why create our own work queue?
>    2.It seems unfortunate to ever dispatch to the main thread. We could probably avoid it by having `WebProcess::ensureNetworkProcessConnection()` set the IPC::Connection::UniqueID of the latest connection to the network process is some global variable (with a global Lock) and use IPC::Connection::send(UniqueID) here directly off the main thread.

That is a great suggestion. I have adopted this approach in the latest patch.